appengine/datastore/src/test/java/com/example/appengine/QueriesTest.java

```diff
@@ -716,21 +716,21 @@ public void queryRestrictions_surprisingMultipleValuesEquals_returnsMatchedEntit
     c.setProperty("x", ImmutableList.<Long>of(-6L, 2L));
     Entity d = new Entity("Widget", "d");
     d.setProperty("x", ImmutableList.<Long>of(-6L, 4L));
-    datastore.put(ImmutableList.<Entity>of(a, b, c, d));
+    Entity e = new Entity("Widget", "e");
+    e.setProperty("x", ImmutableList.<Long>of(1L, 2L, 3L));
+    datastore.put(ImmutableList.<Entity>of(a, b, c, d, e));
 
     // [START surprising_behavior_example_2]
     Query q =
         new Query("Widget")
-            .setFilter(new FilterPredicate("x", FilterOperator.EQUAL, 1))
-            .setFilter(new FilterPredicate("x", FilterOperator.EQUAL, 2));
+            .setFilter(
+                CompositeFilterOperator.and(
+                    new FilterPredicate("x", FilterOperator.EQUAL, 1),
+                    new FilterPredicate("x", FilterOperator.EQUAL, 2)));
     // [END surprising_behavior_example_2]
 
     List<Entity> results = datastore.prepare(q).asList(FetchOptions.Builder.withDefaults());
-    assertThat(getKeys(results)).named("query result keys").contains(a.getKey());
-
-    // Note: When run in the test server, this matches "c" as expected and does
-    // not match "d" as expected.  For some reason it does *not* match "b".
-    // The behavior of queries on repeated values is definitely surprising.
+    assertThat(getKeys(results)).named("query result keys").containsExactly(a.getKey(), e.getKey());
```

Comment:
Yes, but if you set the `setKeysOnly` on the query the returned Entities will only contain Keys (and could be compared as is). I think in `gcloud-java-datastore` we did a better job and return a `List<Key>` rather than a `List<Entity>` which only holds the Keys.